sqlx-postgres/src/copy.rs

```diff
@@ -358,3 +362,192 @@ async fn pg_begin_copy_out<'c, C: DerefMut<Target = PgConnection> + Send + 'c>(
 
     Ok(Box::pin(stream))
 }
+
+enum PgCopyBothCommand {
+    Begin(Vec<u8>),
+    CopyData(Vec<u8>),
+    CopyDone { from_client: bool },
+}
+
+pub struct PgCopyBothSender(flume::Sender<PgCopyBothCommand>);
+pub struct PgCopyBothReceiver(flume::r#async::RecvStream<'static, Result<Bytes>>);
+
+// Open a duplex connection allowing high-speed bulk data transfer to and from the server.
+//
+// # Example
+//
+// ```rust,no_run
+// # use sqlx_postgres::{
+// #     pg_copy_both, PgConnectOptions, PgPoolOptions, PgReplicationMode,
+// # };
+// // Connection must be configured with a replication mode!
+// let options = PgConnectOptions::new()
+//     .host("0.0.0.0")
+//     .replication_mode(PgReplicationMode::Logical);
+//
+// let pool = PgPoolOptions::new()
+//     .connect_with(options)
+//     .await
+//     .expect("failed to connect to postgres");
+//
+// let query = format!(
+//     r#"START_REPLICATION SLOT "{}" LOGICAL {} ("proto_version" '1', "publication_names" '{}')"#,
+//     "test_slot", "0/1573178", "test_publication",
+// );
+// let (copy_tx, copy_recv) = pg_copy_both(&pool, query.as_str())
+//     .await
+//     .expect("start replication");
+// // Read data from the server.
+// while let Some(data) = copy_recv.next().await {
+//     println!("data: {:?}", data);
+//     // And send some back (e.g. keepalive).
+//     copy_tx.send(Vec::new()).await?;
+// }
+// // Connection closed.
+// ```
+pub async fn pg_copy_both(
+    pool: &Pool<Postgres>,
+    statement: &str,
+) -> Result<(PgCopyBothSender, PgCopyBothReceiver), Error> {
```

Comment:
In what way do we give up a lot of control with this approach?
Having two handle seems more flexible (user can group them back into a structure if needed).
+
I thought that it was nice to have a "channel-like" API here as you have the upstream stream of command and the downstream stream of data.

---

If I go the single-object way I either need to have public fields of an `into_sth` to split it it two, right?

Reply:
For now I've made a structure with two public field (so that the caller can deconstruct if needed).
Does it sounds good to you or you had something else in mind?